servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/Host.java

```diff
@@ -15,583 +15,77 @@
-    }
-
-    @Override
-    public Completable onClosing() {
-        return closeable.onClosing();
-    }
-
-    @SuppressWarnings("unchecked")
-    private Completable doClose(final Function<? super C, Completable> closeFunction) {
-        return Completable.defer(() -> {
-            final ConnState oldState = closeConnState();
-            cancelIfHealthCheck(oldState);
-            final Object[] connections = oldState.connections;
-            return (connections.length == 0 ? completed() :
-                    from(connections).flatMapCompletableDelayError(conn -> closeFunction.apply((C) conn)))
-                    .shareContextOnSubscribe();
-        });
-    }
-
-    private void cancelIfHealthCheck(ConnState connState) {
-        if (Host.isUnhealthy(connState)) {
-            @SuppressWarnings("unchecked")
-            HealthCheck<Addr, C> healthCheck = (HealthCheck<Addr, C>) connState.state;
-            LOGGER.debug("{}: health check cancelled for {}.", lbDescription, healthCheck.host);
-            healthCheck.cancel();
-        }
-    }
-
-    @Override
-    public String toString() {
-        final ConnState connState = this.connState;
-        return "Host{" +
-                "lbDescription=" + lbDescription +
-                ", address=" + address +
-                ", state=" + connState.state +
-                ", #connections=" + connState.connections.length +
-                '}';
-    }
-
-    private static final class ActiveState {
-        private final int failedConnections;
-
-        ActiveState() {
-            this(0);
-        }
-
-        private ActiveState(int failedConnections) {
-            this.failedConnections = failedConnections;
-        }
-
-        ActiveState forNextFailedConnection() {
-            return new ActiveState(addWithOverflowProtection(this.failedConnections, 1));
-        }
-
-        @Override
-        public String toString() {
-            return "ACTIVE(failedConnections=" + failedConnections + ')';
-        }
-    }
-
-    private static final class HealthCheck<ResolvedAddress, C extends LoadBalancedConnection>
-            extends DelayedCancellable {
-        private final ConnectionFactory<ResolvedAddress, ? extends C> connectionFactory;
-        private final Host<ResolvedAddress, C> host;
-        private final Throwable lastError;
+    /**
+     * The address of the host
+     */
+    ResolvedAddress address();
 
-        private HealthCheck(final ConnectionFactory<ResolvedAddress, ? extends C> connectionFactory,
-                            final Host<ResolvedAddress, C> host, final Throwable lastError) {
-            this.connectionFactory = connectionFactory;
-            this.host = host;
-            this.lastError = lastError;
-        }
+    /**
+     * A snapshot of the current connection pool.
+     * This is only for debugging purposes and shouldn't be used for implementation details.
+     */
+    Map.Entry<ResolvedAddress, List<C>> asEntry();
```

Comment:
Should this part of the interface? It can lead to bad habits.
I am fine with some duplication and having protected accessors in the implementations.